Fix package manifest validation and typing

rekmarks opened the pull request:
This PR improves the validation and type checking of package manifest objects. Previously, we were casting package manifest objects and using `Partial` needlessly. Now, the `validatePackageManifest` function acts as a type guard by returning the unmodified package manifest object, but with correct types.

As part of this improved validation, `readJsonFile` has been renamed to `readJsonObjectFile` and rewritten to error if the parsed JSON is anything other than a plain object, as opposed to just a falsy value.

rekmarks committed:
Fix package manifest typing

rekmarks committed:
Fix readJsonObjectFile test

Gudahtt commented on src/package-operations.ts:
```diff
@@ -36,6 +36,11 @@ export interface PackageManifest
   readonly [FieldNames.Workspaces]?: string[];
 }
 
+export interface ValidatedPackageManifest extends PackageManifest {
```
Shouldn't this be something more like "MonorepoPackageManifest"? These properties might be absent on a valid manifest for a non-monorepo.

rekmarks replied:
Done in: b46666d

Gudahtt commented on src/package-operations.ts:
```diff
@@ -313,41 +321,42 @@ function getUpdatedDependencyField(
-export function validatePackageManifest(
-  manifest: Partial<PackageManifest>,
+export function validatePackageManifest<
+  T extends Partial<PackageManifest>,
```
I remain a bit confused about why these generic parameters are necessary. It seems that there are two types of manifests that we want to validate: a monorepo manifest, and a non-monorepo manifest. Why do we need _two_ generic parameters, and to pass in which fields to validate? This all seems a bit... overly complicated? Maybe I'm missing something

rekmarks replied:
It was complicated by the fact that there's effectively _three_ different types of package manifests, going from subset to superset.

The first is the root manifest, which could be either a normal or a monorepo `package.json` file. There, we only need to validate the `version`. For the polyrepo case, we also need to validate `name`. For the monorepo case, we need to also validate `private` and `workspaces`. On top of that, deeper in the monorepo workflow, we need to validate `version` and `name` for the manifests of the monorepo's packages.

Rather than keeping a single generic function around, I split the validation logic out of `getPackageManifest` and created separate validation function for all the different cases. Most of these new functions use a single generic, but it's a convenient to take in e.g. a `Partial` manifest object and return it with e.g. one field correctly type.

I think this is probably better, because the generic validation function kind of sucked.

Gudahtt replied:
Part of what I was missing is that you were aiming to perform the absolute minimum validation necessary. I would have probably used just a single validation function that validated `name` and `version`, and also conditionally validated `workspaces` and `private` if they existed (and that `private` was `true` if `workspaces` was present). That'd give us one `PackageManifest` type. Then a simple check that `workspaces` existed would give us a second type, `MonorepoPackageManifest`.

But skipping the name validation for the root does make that a bit more challenging.

rekmarks replied:
> But skipping the name validation for the root does make that a bit more challenging.

Yeah, validating the `name` would require that we mandate that monorepo root manifests have a placeholder name field, which seems kind of bad.

rekmarks committed:
Split validation logic into multiple functions

rekmarks commented on src/utils.test.ts:
```diff
@@ -113,18 +113,22 @@ describe('readJsonFile', () => {
+    );
   });
 
   it('throws an error if the file parses to a falsy value', async () => {
```
Done in: b46666d

Gudahtt commented on src/update.ts:
```diff
@@ -110,7 +119,7 @@ async function updateMonorepo(
   newVersion: string,
   versionDiff: SemverReleaseType,
   rootManifest: Pick<
```
Could this use the `MonorepoPackageManifest` type directly instead, rather than picking just those two properties?

rekmarks committed:
Use type guards, fix types

rekmarks commented:
I think I've addressed all outstanding comments in 51be3ed. The type guards are more verbose, but we avoid casting using `as`, and I think end up in a safer place. I also fixed the various package manifest types.

Gudahtt reviewed (approved):
LGTM!
